File: solution/200.number-of-islands/breadth-first-traversal.ts
/*
 * @lc app=leetcode.cn id=200 lang=typescript
 *
 * https://leetcode-cn.com/problems/number-of-islands
 */

// @lc code=start
function clearIsland(grid: string[][], row: number, col: number) {
    // store unhandled elements of the island
    // element: [rowIndex, colIndex]
    const queue: [number, number][] = [];
    // clear grid element and push it to queue
    const handle = (row: number, col: number) => {
        grid[row][col] = '0';
        queue.push([row, col]);
    };
    handle(row, col);
    while (queue.length > 0) {
        const rc = queue.shift()!;
        const r = rc[0];
        const c = rc[1];
        // up
        if (r - 1 >= 0 && grid[r - 1][c] === '1') {
            handle(r - 1, c);
        }
        // down
        if (r + 1 < grid.length && grid[r + 1][c] === '1') {
            handle(r + 1, c);
        }
        // left
        if (c - 1 >= 0 && grid[r][c - 1] === '1') {
            handle(r, c - 1);
        }
        // right
        if (c + 1 < grid[0].length && grid[r][c + 1] === '1') {
            handle(r, c + 1);
        }
    }
}

function numIslands(grid: string[][]): number {
    if (grid.length === 0 || grid[0].length === 0) {
        return 0;
    }
    let islandNum = 0;
    for (let row = 0; row < grid.length; ++row) {
        for (let col = 0; col < grid[0].length; ++col) {
            if (grid[row][col] === '1') {
                islandNum += 1;
                clearIsland(grid, row, col);
            }
        }
    }
    return islandNum;
}
// @lc code=end

export {};

File: solutions/200.number-of-islands/src/breadth_first_traversal.rs
/*
 * @lc app=leetcode.cn id=200 lang=rust
 *
 * https://leetcode.cn/problems/number-of-islands/
 */

pub struct Solution;

// @lc code=start
use std::collections::VecDeque;

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let rows = grid.len();
        let cols = grid[0].len();
        let mut queue = VecDeque::new();
        let mut count = 0;
        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == '0' {
                    continue;
                }
                queue.push_back((i, j));
                grid[i][j] = '0';
                while !queue.is_empty() {
                    let (i, j) = queue.pop_front().unwrap();
                    if i > 0 && grid[i - 1][j] == '1' {
                        queue.push_back((i - 1, j));
                        grid[i - 1][j] = '0';
                    }
                    if i < rows - 1 && grid[i + 1][j] == '1' {
                        queue.push_back((i + 1, j));
                        grid[i + 1][j] = '0';
                    }
                    if j > 0 && grid[i][j - 1] == '1' {
                        queue.push_back((i, j - 1));
                        grid[i][j - 1] = '0';
                    }
                    if j < cols - 1 && grid[i][j + 1] == '1' {
                        queue.push_back((i, j + 1));
                        grid[i][j + 1] = '0';
                    }
                }
                count += 1;
            }
        }
        count
    }
}
// @lc code=end
